### DISCUSSION TITLE
VB.net Operator '\' in C#

### DISCUSSION THREAD

**Discussion by hiteshdavey** [dotnet/csharplang#4983] [Language Ideas]
There is no VB.net equivalent of '\' oprator. We were in the process of moving our codebase from vb.net to C# and we could not find '\' operator in C#. In very many places of the project, we have extensively used operator '\' for computation.

There is a workaround solution in C# but that is not helping out since this operator is used in many places of the project.

Divides one operand by another and returns an integer result. The division of two numbers returns an integer value. (Decimal part dropped). Example...
a) 123.45 \ 1 will result in 123
b) 501 \ 2 will result in 250

**Comment by bernd5** [comment]
In csharp we have the `/` operator - but not `\` operator which behaves like you describe (just not dropping decimal part implicitly).

[sample](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChis8BOACgFYAGPAAgHoGcBKAblIsstgTZ4cAZgB0AFmrMGeTkA):
```cs
using System;

Console.WriteLine(501 / 2);
Console.WriteLine((int)123.45 / 1);
```

**Reply by hiteshdavey** [reply]
I know how to do this in C#. Since VB.net is now retired from further development; many VB.net developers will try to adopt C# for future development or may decide to change VB app codebase to C# and in such a scenario it will become critical for in-place solution for arithmetic operator "\" 

So far C# has not implemented this "\"  operator, it would be a good idea to add this in C# 10 or C# v.next so all vb.net expressions using this "\" operator make sense when converted to C# else one has to edit the expression and typecast the result which may result in a logical error.

Therefore I suggest the C# design team consider this suggestion.

**Reply by CyrusNajmabadi** [reply]
Could this not just be provided as an extension method?

**Reply by hiteshdavey** [reply]
Yes, there is more than one way to do this; but nothing is more meaningful than having this operator in-built in the C# language.

Writing an extension method or typecasting the result is useful if you need to refactor this change in a couple of places in the codebase but having this operator in-built has more meaning to it. So think about this (after all so many cosmetics changes implemented in C# aggresively; I wonder how come this got a miss in C# language ver 1.0)

**Reply by svick** [reply]
> it would be a good idea to add this in C# 10 or C# v.next so all vb.net expressions using this "\\" operator make sense when converted to C# else one has to edit the expression and typecast the result which may result in a logical error.

There are many differences between VB and C#, I don't think adding this one operator would make porting meaningfully easier. To help you with porting, there are several VB to C# converters, [this one](https://codeconverter.icsharpcode.net/) seems to handle `\` correctly.

> after all so many cosmetics changes implemented in C# aggresively; I wonder how come this got a miss in C# language ver 1.0

Every cosmetic change has to have a justification, and I don't believe "make porting from VB to C# slightly easier" was ever considered a valid reason.

**Comment by HaloFour** [comment]
I don't think it's necessary for C# to add a second operator here.  The only reason VB.NET has both `/` and `\` division operators is because the former always uses floating point division even when the operands are integers.  Both of those operators cause the operands to be implicitly cast to `System.Double` or `System.Long` respectively.  This is unnecessary in C# given that the single `/` operator does either integral or floating point division based on the operands.

IMO, I don't like the behavior of the VB.NET `/` operator.  It's not at all clear that it rounds the operands _prior_ to performing the division and I think that would surprise C# developers.  It only does so for historical reasons with much older dialects of BASIC.  VB.NET devs are going to have to contend with the many other ways that the languages differ from one another and I don't think that porting one operator is going to help, especially when the existing division operator in C# isn't going to be changed to mirror the behavior from VB.NET.
